### PR TITLE
PDF downloading fixes

### LINKED ISSUE #2517: Can't open EPUB/PDF from Project Gutenberg in external app
**Describe the bug**
Clicking on an epub/pdf fails to open it in an external application, instead showing a toast "Could not find an installed application for this type of file"

**Expected behavior**
Epubs/pdfs should be openable in other reader apps

Also of note is this may be due to the deletion of the contentprovider, this may need to be reintroduced explicitly for epub/pdf files

### LINKED ISSUE #2518: ServiceWorker does not support opening epubs/pdfs in external apps
**Describe the bug**
Our service worker client needs to share the capability to open epubs/pdfs that already exists in our WebViewClient

**Expected behavior**
Click a pdf/epub link in bouquineux and have it open in an external application

### LINKED ISSUE #2684: Unable to open PDF files in zimit generated files
From [openzim/zimit/issues/101](../../../openzim/zimit/issues/101):

> I ran a zimit of [survivor library](http://www.survivorlibrary.com/index.php/8-category/43-library-engineering-electrical) ([task here](https://farm.youzim.it/pipeline/1e3882e7a0fe71d29367b016)) - the webpage is a collection of PDF files and has no other fancy content. 
Capture d’écran 2021-08-10 à 09 36 05
> I ended up with a healthy 11.6 Gb file, yet I am unable to open any of the PDF (as in: nothing happens - I don't even get an error message). 

Files can be open / read on Kiwix-serve, so there must be an issue with Android somehow.

### LINKED ISSUE #2709: PDF files accessed from Kiwix fail to open in external PDF reader apps
**Describe the bug**

Opening a PDF from a ZIM in Kiwix opens the dialog to select an application. No matter what application is selected, the application will fail to open the PDF.
**Expected behavior**

PDF files included in ZIMs should be able to be either saved to the device and have a URI provided through a Content Provider, or streamed to the application to prevent duplicate use of storage. At the very least, PDFs accessed through Kiwix should be accessible to the selected external application.
**Steps to reproduce the behavior:**
1. 

1. Access a ZIM that links to an included PDF file.
2. Tap the link to the PDF. Select any PDF application from the dialog.
3. PDF application will open, and fail to load the selected PDF.
**Screenshots**

**Environment**
- Version of Kiwix Android : 
- Device : 
- OS version : 

**Logs**

### LINKED ISSUE #2788: Unable to open Nautilus-based zim files (with PDFs)
There are several files that have been generated via nautilus and consist of PDF files (see, e.g. [this](https://library.kiwix.org/zimgit-post-disaster_en_2022-01) or [that](https://library.kiwix.org/mesptitesquestions_fr_2022-01)).

When trying to open these files I always get an error message "unable to open". Seems a regression to me as it used to work (or even did when I opened [this ticket](../../kiwix-desktop/issues/756))

### PR DESCRIPTION
**Pull request by MohitMaliFtechiz**
Fixes #2709
Fixes #2788
Fixes #2684
Fixes #2518
Fixes #2517

Fixing pdf downloading and rendering issue

### CODE REVIEW

**Comment by MohitMaliFtechiz** [comment]
hi @gouri-panda , can you please this PR too.

File: core/src/main/java/org/kiwix/kiwixmobile/core/main/KiwixWebView.kt [hunk 1]
```diff
@@ -147,37 +144,17 @@ open class KiwixWebView @SuppressLint("SetJavaScriptEnabled") constructor(
...
-            fileToSave.outputStream().use { inputStream.copyTo(it) }
-          }
-          instance.toast(instance.getString(R.string.save_media_saved, fileToSave.name))
-        } catch (e: IOException) {
-          Log.w("kiwix", "Couldn't save image", e)
+        val savedFile = FileUtils.downloadFileFromUrl(url, src, zimReaderContainer)
+        savedFile?.let {
+          instance.toast(instance.getString(R.string.save_media_saved, it.name))
+        } ?: run {
+          Log.w("kiwix", "Couldn't save file")
```
Comment (gouri-panda) [inline core/src/main/java/org/kiwix/kiwixmobile/core/main/KiwixWebView.kt:157]: We are already doing this in `downoloadFileUrl`, so it's unnecessary here.

File: core/src/main/java/org/kiwix/kiwixmobile/core/utils/files/FileUtils.kt [hunk 2]
```diff
@@ -288,4 +288,39 @@ object FileUtils {
...
+  private fun getDecodedFileName(url: String?, src: String?): String =
+    url?.substringAfterLast("/", "")
+      ?.takeIf { it.contains(".") }
+      ?: src?.substringAfterLast("/", "")
+        ?.substringAfterLast("%3A") ?: ""
+
+  @JvmStatic fun downloadFileFromUrl(
+    url: String?,
+    src: String?,
+    zimReaderContainer: ZimReaderContainer
```
Comment (gouri-panda) [inline core/src/main/java/org/kiwix/kiwixmobile/core/utils/files/FileUtils.kt:301]: Nice! Moved to utility class 👍

File: core/src/main/java/org/kiwix/kiwixmobile/core/main/CoreWebViewClient.java [hunk 3]
```diff
@@ -97,11 +101,21 @@ private String convertLegacyUrl(String url) {
...
+      File savedFile = FileUtils.downloadFileFromUrl(url, null, zimReaderContainer);
+      if (savedFile != null && savedFile.exists()) {
+        Uri uri = Build.VERSION.SDK_INT >= Build.VERSION_CODES.N
+          ? FileProvider.getUriForFile(
+          CoreApp.getInstance(),
+          CoreApp.getInstance().getPackageName() + ".fileprovider", savedFile)
+          : Uri.fromFile(savedFile);
+        Intent intent = new Intent(Intent.ACTION_VIEW);
+        intent.setDataAndType(uri, DOCUMENT_TYPES.get(extension));
+        intent.setFlags(Intent.FLAG_ACTIVITY_NO_HISTORY);
```
Comment (gouri-panda) [inline core/src/main/java/org/kiwix/kiwixmobile/core/main/CoreWebViewClient.java:113]: [nit] Use kotlin's `apply` operator.

File: core/src/main/java/org/kiwix/kiwixmobile/core/main/CoreWebViewClient.java [hunk 4]
```diff
@@ -97,11 +101,21 @@ private String convertLegacyUrl(String url) {
...
-      Intent intent = new Intent(Intent.ACTION_VIEW);
-      Uri uri = Uri.parse(url);
-      intent.setDataAndType(uri, DOCUMENT_TYPES.get(extension));
-      intent.setFlags(Intent.FLAG_ACTIVITY_NO_HISTORY);
-      callback.openExternalUrl(intent);
+      File savedFile = FileUtils.downloadFileFromUrl(url, null, zimReaderContainer);
+      if (savedFile != null && savedFile.exists()) {
+        Uri uri = Build.VERSION.SDK_INT >= Build.VERSION_CODES.N
+          ? FileProvider.getUriForFile(
+          CoreApp.getInstance(),
```
Comment (gouri-panda) [inline core/src/main/java/org/kiwix/kiwixmobile/core/main/CoreWebViewClient.java:108]: [nit] We use `CoreApp.getInstance()` multiple times. We should declare a variable here.

File: core/src/main/java/org/kiwix/kiwixmobile/core/main/CoreWebViewClient.java (hunk 3 above)
Comment (MohitMaliFtechiz) [reply core/src/main/java/org/kiwix/kiwixmobile/core/main/CoreWebViewClient.java:113]: its java file.

File: core/src/main/java/org/kiwix/kiwixmobile/core/main/CoreWebViewClient.java (hunk 4 above)
Comment (MohitMaliFtechiz) [reply core/src/main/java/org/kiwix/kiwixmobile/core/main/CoreWebViewClient.java:108]: ok

File: core/src/main/java/org/kiwix/kiwixmobile/core/main/CoreWebViewClient.java (hunk 3 above)
Comment (gouri-panda) [reply core/src/main/java/org/kiwix/kiwixmobile/core/main/CoreWebViewClient.java:113]: D'oh! We still have Java files :)

File: core/src/main/java/org/kiwix/kiwixmobile/core/main/KiwixWebView.kt (hunk 1 above)
Comment (MohitMaliFtechiz) [reply core/src/main/java/org/kiwix/kiwixmobile/core/main/KiwixWebView.kt]: done

**Review by gouri-panda — DISMISSED** [review]
🎉

File: core/src/main/java/org/kiwix/kiwixmobile/core/utils/files/FileUtils.kt [hunk 5]
```diff
@@ -288,4 +288,39 @@ object FileUtils {
     }
     return null
   }
+
+  private fun getDecodedFileName(url: String?, src: String?): String =
```
Comment (kelson42) [inline core/src/main/java/org/kiwix/kiwixmobile/core/utils/files/FileUtils.kt:292]: Can we have a comment explaining does this function? It should be as well unit trsted. Please use standart library to parse URL and URL decode it. Here the implementation looks like really weak.

**Review by kelson42 — CHANGES_REQUESTED** [review]

File: core/src/main/java/org/kiwix/kiwixmobile/core/utils/files/FileUtils.kt (hunk 5 above)
Comment (gouri-panda) [reply core/src/main/java/org/kiwix/kiwixmobile/core/utils/files/FileUtils.kt:292]: @kelson42 I created this function. I should implement the docs and tests.

**Comment by kelson42** [comment]
@MohitMaliFtechiz Please rebase branch

**Comment by MohitMaliFtechiz** [comment]
> @MohitMaliFtechiz Please rebase branch

done

**Comment by MohitMaliFtechiz** [comment]
hi @kelson42 ,

Please merge this branch re basing done.

**Comment by kelson42** [comment]
@MohitMaliFtechiz For the monent there is many things I'm not happy with this PR. Not sure how I should handle the problem.

**Review by kelson42 — APPROVED** [review]
@MohitMaliFtechiz Was not able to test your PR against all the scenarios listed in open tickets. I assume you and @gouri-panda did it rigourously. For the rest, please open a ticket to list the problem with talked about around the code of this PR.

**Comment by MohitMaliFtechiz** [comment]
https://github.com/kiwix/kiwix-android/issues/2875

**Pull request merged**